contracts/src/L2/L2EjectionController.sol

```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.13;
+
+import {IPermissionedRegistry} from "../common/IPermissionedRegistry.sol";
+import {EjectionController} from "../common/EjectionController.sol";
+import {NameUtils} from "../common/NameUtils.sol";
+import {IRegistry} from "../common/IRegistry.sol";  
+
+/**
+ * @title L2EjectionController
+ * @dev L2 contract for ejection controller that facilitates migrations of names
+ * between L1 and L2, as well as handling renewals.
+ */
+abstract contract L2EjectionController is EjectionController {
+    error NotTokenOwner(uint256 tokenId);
+    error InvalidLabel(uint256 tokenId, string label);
+
+    constructor(IPermissionedRegistry _registry) EjectionController(_registry) {}
+
+    /**
+     * @dev Should be called when a name is being migrated back to L2.
+     *
+     * @param tokenId The token ID of the name being migrated
+     * @param transferData The transfer data for the name being migrated
+     */
+    function _completeMigrationFromL1(
+        uint256 tokenId,
+        TransferData memory transferData
+    ) internal virtual {
+        if (registry.ownerOf(tokenId) != address(this)) {
+            revert NotTokenOwner(tokenId);
+        }
+
+        registry.setSubregistry(tokenId, IRegistry(transferData.newSubregistry));
+        registry.setResolver(tokenId, transferData.newResolver);
+        registry.safeTransferFrom(address(this), transferData.newOwner, tokenId, 1, "");
```

Comment:
Thought for a future PR - should we offer a method on the registry that does all of this?

Reply:
Not sure what the use case will be aside from migration. It's essentially a "transfer to someone new with these values set".